tests/infer/test_map.py

```diff
@@ -0,0 +1,75 @@
+import pytest
+import torch
+from torch.autograd import Variable
+
+import pyro
+import pyro.distributions as dist
+from pyro.infer.map import MAP
+from pyro.optim import Adam
+from pyro.util import ng_ones, ng_zeros
+from tests.common import assert_equal
+
+
+def test_map_sample():
+
+    def model(data):
+        mu = pyro.sample("mu", dist.normal, ng_zeros(1), ng_ones(1))
+        for i, x in enumerate(data):
+            pyro.observe("x_{}".format(i), dist.normal, x, mu, ng_ones(1))
+
+    data = Variable(torch.Tensor([1, 3, 4]))
+    optim = Adam({"lr": 1e-1})
+    infer = MAP(model, optim)
+    for _ in range(100):
+        infer.step(data)
+
+    mu = pyro.param("mu")
+    assert_equal(mu.data[0], 2.0, prec=0.1)
+
+
+def test_map_param():
+
+    def model(data):
+        mu = pyro.param("mu", Variable(torch.zeros(1), requires_grad=True))
+        for i, x in enumerate(data):
+            pyro.observe("x_{}".format(i), dist.normal, x, mu, ng_ones(1))
+
+    data = Variable(torch.Tensor([0, 1, 3, 4]))
+    optim = Adam({"lr": 1e-1})
+    infer = MAP(model, optim)
+    for _ in range(100):
+        infer.step(data)
+
+    mu = pyro.param("mu")
+    assert_equal(mu.data[0], 2.0, prec=0.1)
+
+
+def test_map_chain_rule():
+
+    def model():
+        x = pyro.sample("x", dist.normal, ng_zeros(1), ng_ones(1))
+        y = pyro.sample("y", dist.normal, x, ng_ones(1))
+        pyro.observe("z", dist.normal, y, ng_ones(1) * 3.0, ng_ones(1))
```

Comment:
Shouldn't the third argument be `ng_ones * 3.0` (observations)?

Reply:
Theoretically, it is reasonable. I think that `p(z=y|x,y)` is likelihood and `p(x,y|z=y)` is posterior. Solving MAP by hand, we get to the system of equations `2y - x - 3 = 0` and `2x - y = 0`, which in turn yields `x = 1` and `y = 2`.

Reply:
@fehiepsi - The third argument to `observe` is the data (followed by the distribution parameters), and I thought the test intended the observed data to be `[3]`. Though you are right that it wouldn't change the solution in this case, since the normal distribution is symmetric.

Reply:
Right, `pyro.observe("x", dist.normal, a, b, c)` is symmetric in `(a,b)`. I was thinking of this `observe` statement as an additional prior that we're putting on `y`, saying that it should be close to 3. The test should produce the same result either way.